optuna/samplers/_nsgaiii.py

```diff
@@ -0,0 +1,637 @@
+
+
+def _generate_default_reference_point(
+    n_objectives: int, dividing_parameter: int = 3
+) -> np.ndarray:
+    """Generates default reference points which are `uniformly` spread on a hyperplane."""
+    reference_points = np.zeros(
+        (
+            _multi_choose(n_objectives, dividing_parameter),
+            n_objectives,
+        )
+    )
+    for i, comb in enumerate(
+        itertools.combinations_with_replacement(range(n_objectives), dividing_parameter)
+    ):
+        for j in comb:
+            reference_points[i, j] += 1.0
+    return reference_points
+
+
+def _filter_inf(population: list[FrozenTrial]) -> np.ndarray:
+    # Collect all objective values.
+    n_objectives = len(population[0].values)
+    objective_matrix = np.zeros((len(population), n_objectives))
+    for i, trial in enumerate(population):
+        objective_matrix[i] = np.array(trial.values, dtype=float)
+
+    mask_posinf = np.isposinf(objective_matrix)
+    mask_neginf = np.isneginf(objective_matrix)
+
+    # Replace +-inf with nan temporary to get max and min
+    objective_matrix[mask_posinf + mask_neginf] = np.nan
+    nadir_point = np.nanmax(objective_matrix, axis=0)
+    ideal_point = np.nanmin(objective_matrix, axis=0)
+    interval = nadir_point - ideal_point
+
+    # TODO(Shinichi) reconsider alternative value for inf.
+    rows_posinf, cols_posinf = np.where(mask_posinf)
+    objective_matrix[rows_posinf, cols_posinf] = (
+        nadir_point[cols_posinf] + _COEF * interval[cols_posinf]
+    )
+    rows_neginf, cols_neginf = np.where(mask_neginf)
+    objective_matrix[rows_neginf, cols_neginf] = (
+        ideal_point[cols_neginf] - _COEF * interval[cols_neginf]
+    )
+
+    return objective_matrix
+
+
+def _normalize_objective_values(objective_matrix: np.ndarray) -> np.ndarray:
+    """Normalizes objective values of population
+
+    An ideal point z* consists of minimums in each axis. Each objective value of population is
+    then subtracted by the ideal point.
+    An extreme point of each axis is (originally) defined as a minimum solution of achievement
+    scalarizing function from the population. After that, intercepts are calculate as intercepts
+    of hyperplane which has all the extreme points on it and used to rescale objective values.
+
+    We adopt weights and achievement scalarizing function(ASF) used in pre-print of the NSGA-III
+    paper (See https://www.egr.msu.edu/~kdeb/papers/k2012009.pdf).
+    """
+    n_objectives = np.shape(objective_matrix)[1]
+    # Subtract ideal point from objective values.
+    objective_matrix -= np.min(objective_matrix, axis=0)
+    # Initialize weights.
+    weights = np.eye(n_objectives)
+    weights[weights == 0] = 1e6
+
+    # Calculate extreme points to normalize objective values.
+    # TODO(Shinichi) Reimplement to reduce time complexity.
+    asf_value = np.max(
+        np.einsum("nm,dm->dnm", objective_matrix, weights),
+        axis=2,
+    )
```

Comment:
I guess this corresponds to eq. 4 of https://www.egr.msu.edu/~kdeb/papers/k2012009.pdf. We might mention it in the code comment, but we can work on it in a follow-up PR.